src/ucp/core/ucp_am.c

```diff
@@ -533,161 +547,197 @@ ucp_am_handler(void *am_arg, void *am_data, size_t am_length,
     ucp_am_hdr_t *hdr   = (ucp_am_hdr_t *)am_data;
-    uint16_t am_id      = hdr->am_hdr.am_id;
+    uint16_t am_id      = hdr->am_id;
 
     return ucp_am_handler_common(worker, hdr + 1, sizeof(*hdr), am_length,
                                  NULL, am_id, am_flags);
 }
 
-static ucp_am_unfinished_t *
-ucp_am_find_unfinished(ucp_worker_h worker, ucp_ep_h ep,
-                       ucp_ep_ext_proto_t *ep_ext,
-                       ucp_am_long_hdr_t *hdr, size_t am_length)
+static UCS_F_ALWAYS_INLINE ucp_am_unfinished_t*
+ucp_am_find_unfinished(ucp_worker_h worker, ucp_ep_ext_proto_t *ep_ext,
+                       uint64_t msg_id)
 {
     ucp_am_unfinished_t *unfinished;
-    /* TODO make this hash table for faster lookup */
+
     ucs_list_for_each(unfinished, &ep_ext->am.started_ams, list) {
-        if (unfinished->msg_id == hdr->msg_id) {
+        if (unfinished->msg_id == msg_id) {
             return unfinished;
         }
     }
 
     return NULL;
 }
 
-static ucs_status_t
-ucp_am_handle_unfinished(ucp_worker_h worker,
-                         ucp_am_unfinished_t *unfinished,
-                         ucp_am_long_hdr_t *long_hdr,
-                         size_t am_length, ucp_ep_h reply_ep)
+static UCS_F_ALWAYS_INLINE ucp_am_unfinished_t*
+ucp_am_alloc_unfinished(ucp_worker_h worker, ucp_ep_ext_proto_t *ep_ext,
+                        uint64_t msg_id)
+{
+    ucp_am_unfinished_t *unfinished;
+
+    unfinished = ucs_malloc(sizeof(ucp_am_unfinished_t), "ucp multi-frag AM");
+    if (ucs_unlikely(unfinished == NULL)) {
+        ucs_error("failed to allocate UCP AM fragment desc");
+        return NULL;
+    }
+    unfinished->msg_id = msg_id;
+    ucs_queue_head_init(&unfinished->mid_rdesc_q);
+    ucs_list_add_head(&ep_ext->am.started_ams, &unfinished->list);
+
+    return unfinished;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_am_handle_unfinished(ucp_worker_h worker, ucp_am_unfinished_t *unfinished,
+                         void* data, size_t length, size_t offset)
 {
     uint16_t am_id;
     ucs_status_t status;
 
-    memcpy(UCS_PTR_BYTE_OFFSET(unfinished->all_data + 1, long_hdr->offset),
-           long_hdr + 1, am_length - sizeof(*long_hdr));
-    unfinished->left -= am_length - sizeof(*long_hdr);
+    memcpy(UCS_PTR_BYTE_OFFSET(unfinished->all_data + 1, offset), data, length);
+
+    unfinished->left -= length;
     if (unfinished->left == 0) {
-        am_id = long_hdr->am_id;
-        status = worker->am_cbs[am_id].cb(worker->am_cbs[am_id].context,
-                                          unfinished->all_data + 1,
-                                          long_hdr->total_size,
-                                          reply_ep,
-                                          UCP_CB_PARAM_FLAG_DATA);
+        ucs_assert(unfinished->all_data != NULL);
+        ucs_assert(unfinished->total_size > 0);
+
+        am_id = unfinished->am_id;
+        if (!ucp_am_id_valid(worker, am_id)) {
```

Comment:
aren't we already checking id before calling ucp_am_handle_unfinished?

Reply:
no, the id is checked for assembled message only. We do not have am_id in the middle fragments

Reply:
can we check id only when fist frag arrives?

Reply:
what to do with other fragments? I thought that it is also better from error handling perspective, we assemlble message at first and then decide what to do with it

Reply:
ok